### PR TITLE
Rewrite the js part to ts

### PR DESCRIPTION
**Pull request by Duddino**
This also fixes a couple of bugs.
Draft because it needs a bit more testing

### CODE REVIEW

File: js/pivx_shield.ts [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+    }
+    this.extsk = enc_extsk;
+  }
+
+  /**
+   * @returns {string} a string that saves the public shield data.
+   * The seed or extended spending key still needs to be provided
+   * if spending authority is needed
+   */
+  save() {
```
Comment (panleone) [inline js/pivx_shield.ts:243]: can remove {string} from @returns since we are on ts

File: js/pivx_shield.ts [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+      (note) =>
+        !this.unspentNotes.some(
+          (note2) => JSON.stringify(note2[0]) === JSON.stringify(note[0]),
+        ),
+    );
+  }
+
+  /**
+   * Remove the Shield Notes that match the nullifiers given in input
+   * @param {string[]} nullifiers - Array of nullifiers
```
Comment (panleone) [inline js/pivx_shield.ts:325]: can remove string[] from JSDoc since we are on ts

File: js/pivx_shield.ts [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+      .flat()
+      .reduce((acc, v) => acc + v.value, 0);
+  }
+
+  /**
+   * Creates a transaction, sending `amount` satoshis to the address
+   * @param {{address: string, amount: number, blockHeight: number, useShieldInputs: boolean, utxos: UTXO[]?, transparentChangeAddress: string?}} target
+   * @returns {{hex: string, spentUTXOs: UTXO[], txid: string}}
+   */
+  async createTransaction({
```
Comment (panleone) [inline js/pivx_shield.ts:357]: use ts instead of specifying types with JSDoc

File: js/pivx_shield.ts [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+   * @param o - options
+   * @param o.seed - array of 32 bytes that represents a random seed.
+   * @param o.extendedSpendingKey - Extended Spending Key.
+   * @param o.extendedFullViewingKey - Full viewing key
+   * @param o.blockHeight - number representing the block height of creation of the wallet
+   * @param o.coinType - number representing the coin type, 1 represents testnet
+   * @param o.accountIndex - index of the account that you want to generate, by default is set to 0
+   * @param o.loadSaplingData - if you want to load sapling parameters on creation, by deafult is set to true
+   */
+  static async create({
```
Comment (panleone) [inline js/pivx_shield.ts:131]: same here I guess

**Review by panleone — CHANGES_REQUESTED** [review]
tACK, working as expected just some small nits

File: js/pivx_shield.ts (hunk 4 above)
Comment (Duddino) [reply js/pivx_shield.ts:131]: These are for commenting not for types

**Pull request merged**